### DISCUSSION TITLE
Use filter with toggle favorite

### DISCUSSION THREAD

**Discussion by maksymvasylyev** [neuefische/questions#601] [Web] [answered]
Hello
I am trying to use toggle favorite function together with filter search function.
Toggle function works and favorite toggle function works.
But new filter is made with starting data base and favorite items disappear
I don`t know how can I make new filter search with already existed favorite items.

function FilterForm() {
  // const [fav, setFav] = useLocalStorageState("fav", { defaultValue: data });
  const [list, setList] = useLocalStorageState("list", { defaultValue: data });

  function handleToggleFavorite(id) {
    const updatedFavCars = list.map((car) => {
      if (car.id === id) {
        return { ...car, isFavorite: !car.isFavorite };
      } else return car;
    });
    setList(updatedFavCars);
  }

  function handleSubmit(event) {
    event.preventDefault();
    //fetch input data
    const formData = new FormData(event.target);
    const inputData = Object.fromEntries(formData);
    //destructing
    const { CountryOfManufacture, BodyType, Fuel, minPrice, maxPrice } =
      inputData;

    setList(
      data.filter(
        (car) =>
          car.CountryOfManufacture === CountryOfManufacture &&
          car.bodyType === BodyType &&
          car.Fuel.includes(Fuel) &&
          between(
            car.Price,
            Number(minPrice) || 1,
            Number(maxPrice) || 90000000000
          )
      )
    );
    event.target.reset();
  }

**Comment by mbosselmann** [comment]
Hi Max,
can you share your current code with us too? That would be great. :)

**Reply by maksymvasylyev** [reply]
https://github.com/maksymvasylyev/capstone-project/pull/13

**Reply by maksymvasylyev** [reply]
Sure, thanks a lot!

**Comment by mbosselmann** [comment, marked as answer]
Hi Max,
I checked your code and to get both functionalities to work like intended you need to adjust some parts of your code:

1. `_app.js`
```jsx
export default function App({ Component, pageProps }) {
  const [cars, setCars] = useLocalStorageState("list", {
    defaultValue: data,
  });

  function handleToggleFavorite(id) {
    const updatedCars = cars.map((car) => {
      if (car.id === id) {
        return { ...car, isFavorite: !car.isFavorite };
      } else return car;
    });
    setCars(updatedCars);
  }

  return (
    <>
      <GlobalStyle />
      <Component
        {...pageProps}
        cars={cars}
        onToggleFavorite={handleToggleFavorite}
      />
    </>
  );
}
```
 - put your data in the `_app.js` to have one single source of truth for your index page and your filter page
 - add a handleToggleFavorite function to your `_app.js` for updating which car is a favorite
 - hand over cars and handleToggleFavorite as onToggleFavorite to your pages via  `Component`
- I used cars as name as I think this is the best name to describe your data :)

2. next step is to update your `FilterPage`
```jsx
import FilterForm from "@/components/FilterForm/FilterForm";

function FilterPage({ cars, onToggleFavorite }) {
  return <FilterForm cars={cars} onToggleFavorite={onToggleFavorite} />;
}

export default FilterPage;
```

- you need to hand over the props cars and onToggleFavorite to your `FilterForm`
- note: `import React from "react";` is not needed and can be removed (in every file)

3. `FilterForm.js`:
```jsx

function FilterForm({ cars, onToggleFavorite }) {
  const [filteredCars, setFilteredCars] = useState(cars);
  
function handleSubmit(event) {
    event.preventDefault();
    //fetch input data
    const formData = new FormData(event.target);
    const inputData = Object.fromEntries(formData);
    //destructing
    const { CountryOfManufacture, BodyType, Fuel, minPrice, maxPrice } =
      inputData;

    setFilteredCars(
      cars.filter(
        (car) =>
          car.CountryOfManufacture === CountryOfManufacture &&
          car.bodyType === BodyType &&
          car.Fuel.includes(Fuel) &&
          between(
            car.Price,
            Number(minPrice) || 1,
            Number(maxPrice) || 90000000000
          )
      )
    );
    event.target.reset();
  }

  function handleToggleFavorite(id) {
    const updatedFavCars = filteredCars.map((car) => {
      if (car.id === id) {
        return { ...car, isFavorite: !car.isFavorite };
      } else return car;
    });
    setFilteredCars(updatedFavCars);
    onToggleFavorite(id);
  }

  return (
    <>
   ...all your other code ;)
      <StyledResetButton type="button" onClick={() => setFilteredCars(cars)}>
        Clean search result
      </StyledResetButton>

      <FilteredCarsList
        list={filteredCars}
        handleToggleFavorite={handleToggleFavorite}
      />
    </>
  );
}
```
- in your `FilterForm` you declare a new state `filteredCars`, the initial value is you cars data
- your `handleSubmit` does not filter the list anymore, but the `filteredCars`
- create a new function called `handleToggleFavorite` and update `filteredCars` and call `onToggleFavorite` (do not forget to hand over the id)
- hand over to your `FIlteredList` the `filteredCars` and `handleToggleFavorite` (not onToggleFavorite)

The changes should do the magic. I hope this does help. :)
Feel free to ask if you have further questions.

**Reply by maksymvasylyev** [reply]
@mbosselmann Hi Mareike
Thank you very much that you spent your evening time to review my code. I really appreciate it.
Unfortunately nothing works now.

By trying to go through the Link to the filter-form page - I get an error:
"TypeError: Cannot read properties of undefined (reading 'length')" in 
components/FilterForm/FilteredCarsList.js (17:12) @ length

By refreshing - this error disappears. But buy pressing on the toggle button - new error is shown:
TypeError: Cannot read properties of undefined (reading 'map') in 
pages/_app.js (11:29) @ map

By refreshing this error disappears and toggle button works.
By trying to use filter - third error is shown:
TypeError: Cannot read properties of undefined (reading 'filter') in 
components/FilterForm/FilterForm.js (92:11) @ filter

I am sorry, but I need some more help with it, if you could find some free time.
Thank you in advance.🙏🏼

**Reply by mbosselmann** [reply]
We looked over the code together and just had to delete the data from the local storage. Now my suggested solution works. :)

**Reply by maksymvasylyev** [reply]
You are the Best!! Thank you very much!!!
